### PR TITLE
Url: refactor getUrlSearchParams()

### PR DESCRIPTION
**Pull request by leeoniya**
this updates `getUrlSearchParams()` to use the modern [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) API. it also adds a cache so when it is called repeatedly for the same querystring, it skips the processing. this saves about 50% of its cost, droppnig it significantly in a streaming profile.

i originally hoped this can be dropped by 100%, but simply reading `window.location.search` is both necessary and really expensive. maybe we can eliminate the need to call `getUrlSearchParams()` on every data update in the future.

unfortunately the `@ts-ignore`s are necessary because the built-in DOM typings are incomplete :(

for some reason Jest is failing locally for me with the existing `.toStrictEqual()`, despite the output serializing to the same thing :shrug: 

image

**before:**

image

**after:**

image

### CODE REVIEW

**Review by torkelo — APPROVED** [review]

**Comment by leeoniya** [comment]
after some deep debugging into Jest and Jasmine comparison functions, it turns out that Jest's variant of `URLSearchParams.getAll()` returns an array that does not have an equivalent `.constructor` property to a regular Array.

the following in chrome devtools: `true`:

```js
(new URLSearchParams('a=b')).getAll('a').constructor === [].constructor;`
```

in jest:

```js
  it('has identical Array constructors', () => {
    expect(new URLSearchParams('a=b').getAll('a').constructor).toStrictEqual([].constructor);
  });
```

image

while we can appease Jest by wrapping `all` in a spread: `[...all]`. i'm going to change the test itself to compare the serialized results directly.

**Comment by leeoniya** [comment]
the remaining failures in `ShareLink.test.tsx` appear to be be due to the cache being used and maybe some state being carried over from the previous tests (panelId=22). each tests works in isolation, but when run together produce the failures. i tried resetting in those tests with `mockLocationHref()` but without much success.

File: packages/grafana-data/src/utils/url.ts [hunk 1]
```diff
@@ -102,31 +102,32 @@ function appendQueryToUrl(url: string, stringToAppend: string) {
...
+
+    const curParsed = new URLSearchParams(curSearch);
+
+    // @ts-ignore
+    for (const key of new Set(curParsed.keys())) {
+      // @ts-ignore
+      let all = curParsed.getAll(key);
+      // @ts-ignore
+      prevParams[key] = all.length === 1 && all[0] === '' ? true : all;
     }
```
Comment (Clarity-89) [inline packages/grafana-data/src/utils/url.ts:125]:
```suggestion
     // @ts-ignore
    for (const [key, value] of curParsed.entries()) {
      prevParams[key] = [...(prevParams[key] || []), value || true]
    }
```
Wouldn't it be faster to get all the entries so then we can skip the `getAll` step? I guess array spread could be quite expensive though, but maybe using smth like `concat` would help 🤔

Comment (leeoniya) [reply packages/grafana-data/src/utils/url.ts:125]: `curParsed.entries()` won't give you unique keys. so you'll get back things like `["a","1"],["a","2"]` for `"a=1&a=2`:

image

Comment (leeoniya) [reply packages/grafana-data/src/utils/url.ts:125]: eh, nvm. sorry 5am here. i'm not sure which version is faster but with a cache it likely makes no practical difference. concat is probably faster than spread, but i dunno vs `Set()` + `getAll()`.

Comment (Clarity-89) [reply packages/grafana-data/src/utils/url.ts:125]: at least `entries` is less lines then `new Set` + `getAll()` 😂

Comment (leeoniya) [reply packages/grafana-data/src/utils/url.ts:125]: touché

File: packages/grafana-data/src/utils/url.ts [hunk 2]
```diff
@@ -102,31 +102,30 @@ function appendQueryToUrl(url: string, stringToAppend: string) {
...
-      params[key] = true;
+  const curSearch = window.location.search.substring(1);
+
+  if (curSearch !== prevSearch) {
+    prevParams = {};
+
+    const curParsed = new URLSearchParams(curSearch);
+
+    // @ts-ignore
+     // @ts-ignore
```
Comment (leeoniya) [inline packages/grafana-data/src/utils/url.ts:120]:
```suggestion
```

**Review by dprokop — APPROVED** [review]

**Pull request closed without merging**